rust/src/dataset/transaction.rs

```diff
@@ -0,0 +1,627 @@
+        transaction_file_path: &str,
+        config: &ManifestWriteConfig,
+    ) -> Result<(Manifest, Vec<Index>)> {
+        // Get the schema and the final fragment list
+        let schema = match self.operation {
+            Operation::Overwrite { ref schema, .. } => schema.clone(),
+            Operation::Merge { ref schema, .. } => schema.clone(),
+            _ => {
+                if let Some(current_manifest) = current_manifest {
+                    current_manifest.schema.clone()
+                } else {
+                    return Err(Error::Internal {
+                        message: "Cannot create a new dataset without a schema".to_string(),
+                    });
+                }
+            }
+        };
+
+        let mut fragment_id = if matches!(self.operation, Operation::Overwrite { .. }) {
+            0
+        } else {
+            current_manifest
+                .and_then(|m| m.max_fragment_id())
+                .map(|id| id + 1)
+                .unwrap_or(0)
+        };
+        let mut final_fragments = Vec::new();
+        let mut final_indices = current_indices;
+
+        let maybe_existing_fragments =
+            current_manifest
+                .map(|m| m.fragments.as_ref())
+                .ok_or_else(|| Error::Internal {
+                    message: format!(
+                        "No current manifest was provided while building manifest for operation {}",
+                        self.operation.name()
+                    ),
+                });
+
+        match &self.operation {
+            Operation::Append { ref fragments } => {
+                final_fragments.extend(maybe_existing_fragments?.clone());
+                final_fragments.extend(Self::fragments_with_ids(
+                    fragments.clone(),
+                    &mut fragment_id,
+                ));
+            }
+            Operation::Delete {
+                ref updated_fragments,
+                ref deleted_fragment_ids,
+                ..
+            } => {
+                // Remove the deleted fragments
+                final_fragments.extend(maybe_existing_fragments?.clone());
+                final_fragments.retain(|f| !deleted_fragment_ids.contains(&f.id));
+                final_fragments.iter_mut().for_each(|f| {
+                    for updated in updated_fragments {
+                        if updated.id == f.id {
+                            *f = updated.clone();
+                        }
+                    }
+                });
+            }
+            Operation::Overwrite { ref fragments, .. } => {
+                final_fragments.extend(Self::fragments_with_ids(
+                    fragments.clone(),
+                    &mut fragment_id,
+                ));
+                final_indices = Vec::new();
+            }
+            Operation::Rewrite {
+                ref new_fragments, ..
+            } => {
+                final_fragments.extend(Self::fragments_with_ids(
```

Comment:
do we need to retain the frags that aren't rewritten?

Reply:
I'll be fixing this as part of the compaction implementation.